qadence/models/qnn.py

```diff
@@ -15,11 +15,85 @@
 from qadence.noise import Noise
-from qadence.types import BackendName, DiffMode, Endianness
+from qadence.types import BackendName, DiffMode, Endianness, ParamDictType
 
 logger = get_logger(__name__)
 
 
+def transform_output(
+    output_scaling: Tensor | nn.Parameter, output_shifting: Tensor | nn.Parameter
+) -> Callable[[Tensor], Tensor]:
+    def transform(outputs: Tensor) -> Tensor:
+        return output_scaling * outputs + output_shifting
+
+    return transform
+
+
+def format_to_dict_fn(
+    inputs: list[sympy.Symbol | str] = [],
+) -> Callable[[Tensor | ParamDictType], ParamDictType]:
+    """Format an input tensor into the format required by the forward pass.
+
+    The tensor is assumed to have dimensions: n_batches x in_features where in_features
+    corresponds to the number of input features of the QNN
+    """
+    in_features = len(inputs)
+
+    def tensor_to_dict(values: Tensor | ParamDictType) -> ParamDictType:
+        if isinstance(values, Tensor):
+            values = values.reshape(-1, 1) if len(values.size()) == 1 else values
+            if not values.shape[1] == in_features:
+                logger.error(f"Model expects in_features={in_features} but got {values.shape[1]}.")
+            values = {fparam.name: values[:, inputs.index(fparam)] for fparam in inputs}  # type: ignore[union-attr]
+        return values
+
+    return tensor_to_dict
+
+
+def transform_input(
+    input_scaling: Tensor | nn.Parameter,
+    input_shifting: Tensor | nn.Parameter,
+    inputs: list[sympy.Symbol | str],
+) -> Callable[[ParamDictType | Tensor], ParamDictType | Tensor]:
+    """
+    Returns a function which scales and shifts the user-provided values for the FeatureParameters.
+
+    Note that 'input_scaling' and 'input_shifting' have to contain values for each FeatureParameter.
+    Arguments:
+        input_scaling: A torch tensor or torch.nn.Parameter for scaling the input.
+        input_shifting: A torch tensor or torch.nn.Parameter for shifting the input.
+        inputs: A list of sympy symbols or strings of the FeatureParameters.
+
+    Returns:
+        A function for scaling and shifting FeatureParameters passed either as a dict or Tensor.
+    """
+    input_scaling, input_shifting = list(
+        map(lambda t: t.reshape(-1, 1) if len(t.shape) < 2 else t, [input_scaling, input_shifting])
+    )
+    format_to_dict = format_to_dict_fn(inputs)
+
+    def transform(values: ParamDictType | Tensor) -> ParamDictType | Tensor:
+        """Scales and shifts user-provided FeatureParameters 'values'.
+
+        'values' has to either be a Tensor when using torch.nn.Module, or a standard values dict.
+
+        Arguments:
+            values: A torch Tensor or a dict containing values for Featureparameters.
+        Returns:
+            A dict containing scaled and shifted FeatureParameters.
+        """
+        if isinstance(values, Tensor):
+            values = format_to_dict(values)
+        values = {
```

Comment:
@smitchaudhary i changed the input scaling/shifting requirements to ALWAYS be 2D